### ISSUE TITLE
About e4e Model

### ISSUE THREAD

**Issue by xuedue** [yzliu567/sc-gan#2]
Hello author, thank you for your work and code, my attribute editing work would like to use your method as one of the comparison methods, since our method does not train the attribute of pose, I wonder if you can provide the model e4e_ffhq_agesmileglasses.pt. Thanks again.

**Comment by yzliu567** [comment]
Hi. Sorry we don't have this model on hand. You can follow `encoder4editing/train_sh.txt` to train one from the given e4e pretrained model and our agesmileglasses GAN model. You may need to modify some codes: [dataset](https://github.com/yzliu567/sc-gan/blob/main/encoder4editing/datasets/images_dataset.py#L20) and [train](https://github.com/yzliu567/sc-gan/blob/main/encoder4editing/training/coach.py#L99). You can find the files you need in `distance` folder(`glassessmileage_min_max.txt`, `smile_LARGE.txt`).

**Comment by xuedue** [comment]
> 

Hello author, sorry to bother you again, it seems that more code needs to be changed in the training process, I would like to ask about the order of the attributes in df_mode, from the inference code, it should be [1:4] to control age, [4:7] to control smile, [7:10] to control glasses, I don't know if there is something wrong with my understanding.

**Comment by yzliu567** [comment]
Hi, the order should be age-glasses-smile. See [this](https://github.com/yzliu567/sc-gan/blob/main/dataset_ffhq.py#L53) for the attribute order. You can also uncomment code [here](https://github.com/yzliu567/sc-gan/blob/main/train_mmd.py#L244) to visualize the image and corresponding mode vector to have a check.

**Comment by xuedue** [comment]
> 
Hello author, I still have questions about this order. In inference.py I set ganpath to agsmileglasses.pt, set ckpt to pre-trained original e4e without finetune. Then edited with conditional_pti.py,  the other parameters are the same as written in the readme except for the path. Test Image is 0.jpg in CelebA-HQ and the output can be seen as follows,
multi_base_00.jpg
![multi_base_00](https://user-images.githubusercontent.com/37701943/203038641-9c63a235-490e-4c43-9539-ba8f6cc04080.jpg)
multi_0_0_00.jpg
![multi_0_0_00](https://user-images.githubusercontent.com/37701943/203038818-81cf32bb-b3d1-403e-a989-8c850876d145.jpg)
multi_1_1_00.jpg
![multi_1_1_00](https://user-images.githubusercontent.com/37701943/203038992-58afa21e-1b9e-4615-bfb5-e783919cc986.jpg)
multi_2_2_00.jpg
![multi_2_2_00](https://user-images.githubusercontent.com/37701943/203039137-9942654b-3ed0-4a15-a5ae-17761efd733b.jpg)
 according to the code speculation mode[:,1:4] should be controlling age, mode[:,4:7] should be controlling smile, mode[:,7:10] should be controlling glasses, I do not know if I am where I misunderstood.

**Comment by yzliu567** [comment]
Sorry for the confusion, you are right.

**Comment by xuedue** [comment]
> Sorry for the confusion, you are right.

Thank you for your patience. Since StyleCLIP is also one of the methods we compare, I would like to ask you which method among optimization, mapper and global direction you choose when comparing StyleCLIP. Do you add pivot tuning mode to StyleCLIP when comparing? And how do you balance the parameters of StyleCLIP with the parameters of your method when comparing? For example, how to set the alpha and beta in the global direction of StyleCLIP to be more fair when comparing? Thanks again and I look forward to your reply!

**Comment by xuedue** [comment]
Another question, how to control the strength and direction of editing when editing properties? In my understanding, is it adjusted by modifying the step_size parameter in conditional_pti.py?

**Comment by rinongal** [comment]
The step_size parameter will control strength when doing multi-attribute editing, yes. 

If you want to make an adjustment to the single-mode sample you'll need to change this line:
https://github.com/yzliu567/sc-gan/blob/610575627fd4732e84df8389722bdafc3531f4ad/pti/conditional_pti.py#L152 so that `torch.linspace(0.0, 1.0, steps=num_samples)` is replaced with a list of the values you want to check.

**Comment by xuedue** [comment]
Hello author, if I want to train an additional attribute such as 'Blad', how should I get the data needed for training? Also, I see that the number of attributes in the code is 3. If I want to train a model that can edit 8 attributes or even more attributes, is it feasible? Besides setting mode=1+3*8 and preparing training data, do I need to change any other codes?

**Comment by yzliu567** [comment]
Yes. You just need all the attributes' values, the uniform sampling result and changing the mode number in the architecture. But too many attributes may not work well since the attributes are not independent(e.g. glasses and age). So uniform sampling in the high dimensional space may cover only a small part of the original dataset.

**Comment by xuedue** [comment]
> Yes. You just need all the attributes' values, the uniform sampling result and changing the mode number in the architecture. But too many attributes may not work well since the attributes are not independent(e.g. glasses and age). So uniform sampling in the high dimensional space may cover only a small part of the original dataset.

Thanks for your patient answer, I'm still a bit confused after looking through the code. Suppose I have now obtained the 8 attribute boundaries of interfacegan, such as greyhair.npy, lipstick.npy and so on. Then how should I obtain the training files of these attributes, such as greyhair_large.txt, greyhair_min_max.txt, lipstick_large.txt, lipstick_min_max.txt, etc. I don't seem to see this script in your code.

**Comment by yzliu567** [comment]
Hello. You need to first invert the original dataset into latent vectors (70000 x 18 x 512) using official e4e model. Then compute the distances to the boundaries you pick (which is a simple dot product). The output is distance, whose shape should be (70000,1). And this is xxx_large.txt. After that, you can use our provided script `sample.py` to get uniform samples(index_xxx.txt), minmax values(xxx_min_max.txt). Note: If you are using multiple attributes, you need to do the uniform sampling on all the attributes together, so you will have only 1 min_max and index file.

**Comment by xuedue** [comment]
> Hello. You need to first invert the original dataset into latent vectors (70000 x 18 x 512) using official e4e model. Then compute the distances to the boundaries you pick (which is a simple dot product). The output is distance, whose shape should be (70000,1). And this is xxx_large.txt. After that, you can use our provided script `sample.py` to get uniform samples(index_xxx.txt), minmax values(xxx_min_max.txt). Note: If you are using multiple attributes, you need to do the uniform sampling on all the attributes together, so you will have only 1 min_max and index file.

Hello author, I use your code train_mmd.py to train on an single RTX3090, but the training time seems to be too long, is this normal?
![微信图片_20221202192920](https://user-images.githubusercontent.com/37701943/205285517-50f8281c-f3a1-4b00-a91a-571ceed3fe34.png)
If g_reg_every is used, I estimate that it takes about 100 days to train. If g_reg_every is not used, I estimate that it takes about 13 days to train.
I would like to ask what graphics card are you training on? How long have you been training?

**Comment by yzliu567** [comment]
We used 1 single Titan Xp to finetune the model for 30k to 50k iterations, which takes about 12 hrs. Please refer to our paper for more details.

**Comment by xuedue** [comment]
> Yes. You just need all the attributes' values, the uniform sampling result and changing the mode number in the architecture. But too many attributes may not work well since the attributes are not independent(e.g. glasses and age). So uniform sampling in the high dimensional space may cover only a small part of the original dataset.

Hello, I followed the settings in your paper, specifically, I first used interfacegan to obtain 8 attribute directions, then used e4e pretrained model to map the ffhq dataset to latent code, and obtained 8 label files (bald_LARGE.txt ', 'gender_LARGE.txt', 'grayhair_LARGE.txt', 'lipstick_LARGE.txt', 'makeup_LARGE.txt', 'narroweye_LARGE.txt', 'beard_LARGE.txt', 'smile_LARGE.txt), then I use sample.py to get min_max.txt and index.txt. When fine-tuning stylegan, I modified the codes of dataloader, mode, and loss. After completing the training of stylegan, I modified the code of mode in e4e and fine-tune e4e pretrained model. After training, I tested the datasets of ffhq and celeba-hq according to your test code.

The test results seem to be unsatisfactory, some of the attributes have not been effectively edited, and the multi-attribute editing may also cause distortion of the face. I tested the boundary of interfacegan, each boundary is able to edit attribute effectively, and I  I double-checked the modified code, it should be no problem. I wonder what might be the reason behind this situation. Is this due to too many attributes? Or is it a matter of selected attributes? I don't know how to explain this phenomenon from your method.

**Comment by yzliu567** [comment]
It is likely caused by too many attributes. As I mentioned above, when the attributes are mutually dependent and the number of attributes is large, the uniform sampling method may only pick a small subset of the original dataset. You can try to check the number of different indices in your sampled index.txt.
